fontir/src/ir.rs

```diff
@@ -975,6 +986,59 @@ impl GlyphBuilder {
         self.sources.insert(unique_location.clone(), source);
         Ok(())
     }
+
+    /// * see <https://github.com/googlefonts/ufo2ft/blob/b3895a96ca910c1764df016bfee4719448cfec4a/Lib/ufo2ft/outlineCompiler.py#L1666-L1694>
+    pub fn new_notdef(
+        default_location: NormalizedLocation,
+        upem: u16,
+        ascender: f32,
+        descender: f32,
+    ) -> Self {
+        let upem = upem as f32;
+        let width: u16 = (upem * 0.5).ot_round();
+        let width = width as f64;
+        let stroke: u16 = (upem * 0.05).ot_round();
+        let stroke = stroke as f64;
+
+        let mut path = BezPath::new();
+
+        // outer box
+        let x_min = stroke;
+        let x_max = width - stroke;
+        let y_max = ascender as f64;
+        let y_min = descender as f64;
+        path.move_to((x_min, y_min));
+        path.line_to((x_max, y_min));
+        path.line_to((x_max, y_max));
+        path.line_to((x_min, y_max));
+        path.line_to((x_min, y_min));
+        path.close_path();
+
+        // inner, cut out, box
+        let x_min = x_min + stroke;
+        let x_max = x_max - stroke;
+        let y_max = y_max - stroke;
+        let y_min = y_min + stroke;
+        path.move_to((x_min, y_min));
+        path.line_to((x_min, y_max));
+        path.line_to((x_max, y_max));
+        path.line_to((x_max, y_min));
+        path.line_to((x_min, y_min));
+        path.close_path();
+
+        Self {
+            name: NOTDEF.clone(),
+            codepoints: HashSet::from([0]),
```

Comment:
I'd double-check with @anthrotype but I'm not sure .notdef has any codepoints. Some fonts include a NUL glyph in addition to notdef, although I'm not sure about (and am curious to learn more!) the semantics/rationale for this.

Reply:
Oswald cmap and glyphorder are now identical so I think this is right. I will double-check with Cosimo just in case it's only _sometimes_ right or not as simplistic as I made it. A lot of these little details turn out to be fiddly.

Reply:
Colin was right, the `codepoints` for the generated .notdef glyph must be left empty, .notdef should be understood as the absence of a character mapping; 0x0000 is a different unicode codepoint denoting the NULL character (like in NULL-terminated C strings), it used to be among the recommended first 4 glyphs in old TrueType but current OT spec no longer does that and modern fonts don't contain NULL (and neither the carriage return CR character), but they all contain .notdef (un-cmapped) glyph as the first one (index=0).